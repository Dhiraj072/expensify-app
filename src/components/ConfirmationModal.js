import React from 'react';
import Modal from 'react-modal';

export default (props) => (
    <Modal
        isOpen={props.isModalOpen}
        contentLabel="Remove expense confirmation"
        closeTimeoutMS={300}
        className="modal"
    >
        <div>
            <p>Remove expense?</p>
            <button
                value="yes"
                className="button button--margin"
                onClick={props.handleConfirmation}
            >
                Yes
            </button>
            <button
                value="no"
                className="button button--margin"
                onClick={props.handleConfirmation}
            >
                No
            </button>
        </div>
    </Modal>
);
