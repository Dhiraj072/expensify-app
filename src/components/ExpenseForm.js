import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
    state = {
        description: this.props.expense ? this.props.expense.description : '',
        amount: this.props.expense ? (this.props.expense.amount / 100).toString() : '',
        note: this.props.expense ? this.props.expense.note : '',
        createdAt: this.props.expense ? moment(this.props.expense.createdAt) : moment(),
        calendarFocused: false,
        error: '',
    };
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };
    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };
    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };
    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.amount || !this.state.description) {
            this.setState(() => ({ error: 'Please enter values for description and amount' }));
        } else {
            this.setState(() => ({ error: undefined }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 10,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note,
            });
        }
    };
    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        type="text"
                        placeholder="amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        isOutsideRange={() => false}
                        numberOfMonths={1}
                    />
                    <textarea
                        placeholder="Add note (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    />
                    <button type="submit">Add Expense</button>
                </form>
            </div>
        );
    }
}
