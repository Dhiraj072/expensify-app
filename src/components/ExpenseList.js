import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpList = (props) => (
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-desktop">Expense</div>
            <div className="show-for-desktop">Amount</div>
        </div>
        <div className="list-body">
            { props.expenses.length > 0 ? (
                props.expenses.map((expense) => <ExpenseListItem expense={expense} key={expense.id} />)
            ) : (
                <div className="list-item list-item__message"><span>No expenses</span></div>
            )}
        </div>
    </div>
);

const mapStateToProps = (state) => ({
    expenses: selectExpenses(state.expenses, state.filters),
});

export default connect(mapStateToProps)(ExpList);
