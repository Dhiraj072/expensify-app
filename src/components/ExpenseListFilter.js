import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';

export class ExpListFilter extends React.Component {
    state = {
        calendarFocused: null,
    }
    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    }
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };
    onSortByChange = (e) => {
        if (e.target.value === 'amount') {
            this.props.sortByAmount();
        } else if (e.target.value === 'date') {
            this.props.sortByDate();
        }
    };
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };
    render() {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input
                            type="text"
                            className="text-input"
                            value={this.props.filters.text}
                            onChange={this.onTextChange}
                        />
                    </div>
                    <div className="input-group__item">
                        <select
                            value={this.props.filters.sortBy}
                            className="select"
                            onChange={this.onSortByChange}
                        >
                            <option value="amount">Amount</option>
                            <option value="date">Date</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            endDate={this.props.filters.endDate}
                            onDatesChange={this.onDatesChange}
                            onFocusChange={this.onFocusChange}
                            focusedInput={this.state.calendarFocused}
                            numberOfMonths={1}
                            showClearDates
                            isOutsideRange={() => false}
                        />
                    </div>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    filters: state.filters,
});

const mapDispatchToProps = (dispatch) => ({
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    sortByAmount: () => dispatch(sortByAmount()),
    sortByDate: () => dispatch(sortByDate()),
    setTextFilter: (text) => dispatch(setTextFilter(text)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpListFilter);
