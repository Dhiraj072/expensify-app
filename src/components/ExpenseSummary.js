import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectExpenseTotal from '../selectors/selectExpenseTotal';


export const ExpSummary = (props) => (
    <div className="page-header">
        <div className="content-container">
            <h2 className="page-header__title">
                Viewing
                <span> {props.numExpenses} </span>
                expenses totalling
                <span> {numeral(props.expenseTotal / 100).format('$0,0.00')}</span>
            </h2>
            <div className="page-header__actions">
                <Link className="button" to="/create">Add Expense</Link>
            </div>
        </div>
    </div>
);

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    return {
        numExpenses: visibleExpenses.length,
        expenseTotal: selectExpenseTotal(visibleExpenses),
    };
};

export default connect(mapStateToProps)(ExpSummary);
