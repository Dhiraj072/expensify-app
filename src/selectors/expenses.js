import moment from 'moment';

// Get visible expenses
const getVisibleExpenses = (expenses, {
    startDate, endDate, text, sortBy,
}) => expenses.filter((expense) => {
    const startDateMatch = startDate ? moment(startDate).isSameOrBefore(expense.createdAt, 'day') : true;
    const endDateMatch = endDate ? moment(endDate).isSameOrAfter(expense.createdAt, 'day') : true;
    const testMatch = expense.description.toLowerCase().includes(text.toLowerCase());
    return startDateMatch && endDateMatch && testMatch;
}).sort((a, b) => {
    if (sortBy === 'date') {
        return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'amount') {
        return a.amount < b.amount ? 1 : -1;
    }
});

export default getVisibleExpenses;
