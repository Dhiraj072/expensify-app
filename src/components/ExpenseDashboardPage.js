import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilter from './ExpenseListFilter';

const ExpenseDashboardPage = () => (
    <div>
        Dashboard
        <ExpenseListFilter />
        <ExpenseList />
    </div>
);

export default ExpenseDashboardPage;
