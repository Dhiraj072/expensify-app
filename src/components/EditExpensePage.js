import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import ConfirmationModal from './ConfirmationModal';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    constructor() {
        super();
        this.state = {
            isModalOpen: false,
        };
    }
    onSubmit = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    };
    onClick = () => {
        this.setState(() => ({ isModalOpen: true }));
    };
    handleRemoveExpenseConfirmation = (e) => {
        e.preventDefault();
        const removeExpense = e.target.value;
        if (removeExpense === 'yes') {
            this.props.startRemoveExpense(this.props.expense.id);
            this.props.history.push('/');
        }
        this.setState(() => ({ isModalOpen: false }));
    };
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm onSubmit={this.onSubmit} expense={this.props.expense} />
                    <button className="button" onClick={this.onClick}>
                        Remove Expense
                    </button>
                </div>
                <ConfirmationModal
                    isModalOpen={this.state.isModalOpen}
                    handleConfirmation={this.handleRemoveExpenseConfirmation}
                />
            </div>
        );
    }
}

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id),
});

const mapDispatchToProps = (dispatch) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (id) => dispatch(startRemoveExpense(id)),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
