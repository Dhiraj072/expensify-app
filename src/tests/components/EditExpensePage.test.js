import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';

let startEditExpense;
let history;
let startRemoveExpense;
let wrapper;
const expense = expenses[0];

beforeEach(() => {
    startEditExpense = jest.fn();
    history = { push: jest.fn() };
    startRemoveExpense = jest.fn();
    wrapper = shallow(<EditExpensePage
        startEditExpense={startEditExpense}
        startRemoveExpense={startRemoveExpense}
        history={history}
        expense={expense}
    />);
});

test('should render EditExpensePage correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should handle edit expense', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expense);
    expect(startEditExpense).toHaveBeenLastCalledWith(expense.id, expense);
    expect(history.push).toHaveBeenLastCalledWith('/');
});

test('should handle remove expense', () => {
    wrapper.find('button').simulate('click');
    expect(wrapper.state('isModalOpen')).toBe(true);
});

test('should handle remove', () => {
    wrapper.instance().handleRemoveExpenseConfirmation({
        preventDefault: () => {},
        target: { value: 'yes' },
    });
    expect(startRemoveExpense).toHaveBeenCalledWith(expense.id);
    expect(history.push).toHaveBeenCalledWith('/');
    expect(wrapper.state('isModalOpen')).toBe(false);
});
