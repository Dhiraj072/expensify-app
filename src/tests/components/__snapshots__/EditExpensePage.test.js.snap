// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 100,
          "createdAt": 0,
          "description": "test expense 1",
          "id": "1",
          "note": "test note 1",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button"
      onClick={[Function]}
    >
      Remove Expense
    </button>
  </div>
  <Component
    handleConfirmation={[Function]}
    isModalOpen={false}
  />
</div>
`;
